internal/pkg/service/buffer/sink/tablesink/storage/repository/file_repo.go

```diff
@@ -46,19 +43,52 @@ func (r *FileRepository) List(parentKey fmt.Stringer) iterator.DefinitionT[stora
 	return r.schema.AllLevels().InObject(parentKey).GetAll(r.client)
 }
 
+func (r *FileRepository) ListInLevel(level storage.Level, parentKey fmt.Stringer) iterator.DefinitionT[storage.File] {
+	return r.schema.InLevel(level).InObject(parentKey).GetAll(r.client)
+}
+
 func (r *FileRepository) Get(k storage.FileKey) op.ForType[*op.KeyValueT[storage.File]] {
 	return r.get(k).WithEmptyResultAsError(func() error {
 		return serviceError.NewResourceNotFoundError("file", k.String(), "sink")
 	})
 }
 
-func (r *FileRepository) Create(sinkKey key.SinkKey, credentials *keboola.FileUploadCredentials) *op.AtomicOp[storage.File] {
+func (r *FileRepository) Create(fileKey storage.FileKey, credentials *keboola.FileUploadCredentials) *op.AtomicOp[storage.File] {
 	var sinkKV *op.KeyValueT[definition.Sink]
+	var oldLocalFiles []storage.File
 	var result storage.File
 	return op.Atomic(r.client, &result).
-		// Sink must exist
-		ReadOp(r.all.sink.Get(sinkKey).WithResultTo(&sinkKV)).
-		// Save
+		// Get sink, it must exist
+		ReadOp(r.all.sink.Get(fileKey.SinkKey).WithResultTo(&sinkKV)).
+		// There can be a maximum of one old file in the storage.FileWriting state,
+		// it is atomically switched to the storage.FileClosing state.
+		ReadOp(r.ListInLevel(storage.LevelLocal, fileKey.SinkKey).WithResultTo(&oldLocalFiles)).
+		WriteOrErr(func() (op.Op, error) {
+			var count int
+			var closeFileOp op.Op
+			for _, oldFile := range oldLocalFiles {
+				if oldFile.FileKey == fileKey {
+					// File already exists
+					return nil, serviceError.NewResourceAlreadyExistsError("file", fileKey.String(), "sink")
+				}
+				if oldFile.State == storage.FileWriting {
+					modified := oldFile
+					if err := modified.StateTransition(fileKey.OpenedAt().Time(), storage.FileClosing); err != nil {
+						return nil, err
+					}
+
+					count++
+					closeFileOp = r.put(oldFile, modified, false)
+				}
+			}
+
+			if count > 1 {
+				return nil, errors.Errorf(`unexpected state, found %d opened files in sink "%s"`, count, fileKey.SinkKey)
+			}
+
+			return closeFileOp, nil
+		}).
```

Comment:
Implementation:
- There can be 0 or 1 old file in the state `FileWriting`.
- The old file is switched to the state `FileClosing`.
- Because, we are creating a new file in the state `FileWriting`.
- It all happens atomically in one transaction.